### PR TITLE
Exceptions thrown while processing sequence ops do not result in container closure

### PR DESCRIPTION
**Pull request by vladsud**
Please see issue https://github.com/microsoft/FluidFramework/issues/7800 for more details
That's not ideal catch-all solution, but I do not see opt-in to be working here well (so chose opt-out model).
If you see better ways to ensure right stuff happens here - please let me know.

Need to add UT :)

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 1]
```diff
@@ -433,9 +503,10 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
      * For messages from a remote client, this will be undefined.
      */
     private process(message: ISequencedDocumentMessage, local: boolean, localOpMetadata: unknown) {
-        this.emit("pre-op", message, local, this);
+        this.verifyNotClosed(); // This will result in container closure.
+        this.safeRaiseEvent("pre-op", message, local, this);
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:507]: Do we have a work item to get rid of pre-op entirely?  I think prosemirror is the only thing still using it?

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 2]
```diff
@@ -126,6 +140,61 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+        throw error;
+    }
+
+    /**
+     * Helper method that can be used to emit events and respond to exceptions thrown from event listeners
+     * by closing this object. Please see eventListenerErrorHandler() & closeWithError() for more info.
+     * @param event - event to raise
+     * @param args - payload of event
+     */
+    protected safeRaiseEvent(event, ...args) {
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:190]: Probably default to making all of these private until we have motivation for protected, easier to open up than to close down the interface.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 3]
```diff
@@ -126,6 +140,61 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+     * Event listener handler helper that can be used to react to exceptions thrown from event listeners
+     * It wraps error with DataProcessingError, closes this object and throws resulting error.
+     * See closeWithError() for more details
+     * Ideally such situation never happens, as consumers of DDS should never throw exceptions
+     * in event listeners (i.e. catch any of the issues and make determination on how to handle it).
+     * When such exceptions propagate through, most likely data model is no longer consistent, i.e.
+     * DDS state does not match what user sees. Because of it DDS moves to "corrupted state" and does not
+     * allow processing of ops or local changes, which very quickly results in container closure.
+     */
+     protected eventListenerErrorHandler(event: string | symbol, e: any) {
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:173]: Nit:
```suggestion
    protected eventListenerErrorHandler(event: string | symbol, e: any) {
```

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 4]
```diff
@@ -126,6 +140,61 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+    protected closeWithError(error: any) {
+        if (this.closeError === undefined) {
+            this.closeError = error;
+        }
+    }
+
+    /**
+     * Verifies that this object is not closed via closeWithError(). If it is, throws an error used to close it.
+     */
+     protected verifyNotClosed() {
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:157]: Nit:
```suggestion
    protected verifyNotClosed() {
```

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 5]
```diff
@@ -102,13 +108,21 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
      * @param attributes - Attributes of the shared object
+     * @param closeOnEventListenerException - if true (default), results in critical error (container closure) on
+     * exception thrown from from any event listeners for any events raised by this object.
      */
     constructor(
         public id: string,
         protected runtime: IFluidDataStoreRuntime,
-        public readonly attributes: IChannelAttributes)
+        public readonly attributes: IChannelAttributes,
+        closeOnEventListenerException = true)
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:118]: I wouldn't even provide an opt-out, if someone screws up "op" processing then we're in danger I think.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 6]
```diff
@@ -433,9 +503,10 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
      * For messages from a remote client, this will be undefined.
      */
     private process(message: ISequencedDocumentMessage, local: boolean, localOpMetadata: unknown) {
-        this.emit("pre-op", message, local, this);
+        this.verifyNotClosed(); // This will result in container closure.
+        this.safeRaiseEvent("pre-op", message, local, this);
         this.processCore(message, local, localOpMetadata);
-        this.emit("op", message, local, this);
+        this.safeRaiseEvent("op", message, local, this);
```
Comment (ChumpChief) [inline packages/dds/shared-object-base/src/sharedObject.ts:509]: A potentially more scoped (and maybe better) option - stop emitting "pre-op" and "op" entirely in `SharedObject`, and only emit in sequence during `processCore` (since most DDS don't need to emit "op")

**Review by ChumpChief — COMMENTED** [review]
My main thought is that we shouldn't be emitting "op" and "pre-op" at all anyway, maybe there's a scoped option that lifts this out of `SharedObject`.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 6 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:509]: I'm not after these events (yes, they would be nice to be removed, but that seems orthogonal problem).
The problem is all the events (TEvent interface and actual usage in derived classes) and listeners throwing exceptions.

We could get rid of EventEmitterWithErrorHandling here and simply propagating exceptions (not eating them - that's what we do) will do the right thing in most cases (at least when processing remote ops).
My worry with this more simple story is that local changes (and exceptions thrown from events due to local changes) would leave DDS in non-consistent state which is very hard to debug down the road. Basically there is no RAII and it's safe to assume that exception anywhere is a game over.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 1 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts]: See my comment below, that's not the issue I'm after (even though I agree with you fully here)

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 6 above)
Comment (ChumpChief) [reply packages/dds/shared-object-base/src/sharedObject.ts]: Oh sorry I failed to notice what your super call was doing.

I don't know about closing the container for just any event listener -- event listeners on DDSs (either directly or indirectly through DOs) will include views etc. which based on my experience tend to throw a bunch of silly errors that don't matter.  I worry a bit that overaggressive nuking will make the platform harder to use than is necessary.

You mention in the linked issue that customers should never throw errors to us -- I feel like that is the more correct solution, as catching the error in the handler gives them better recovery options (e.g. kill off a single handler rather than letting the whole container die).  But I do respect the fail-fast strategy, so I wouldn't be strictly opposed.

To minimized confused bug reports on why the container is closed, is there a way to include some text in the `DataProcessingError` like "You threw an error to a SharedObject on event handler 'foo', don't do that, we're closing the container as a result"?  Ideally something that shows up in the error console too I think.

Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts]: Yep, that's the goal - to force customers not to throw at us :) So we need some way to expose such errors to them - silently eating them is not an option IMHO. We can always build some API (containers warnings is that direction, though it has a lot of issues that could be probably improved, if we want to go that route). I prefer fail-fast approach - it helps us find our bugs sooner and it also leaves a way for customers to do right thing.

RE DataProcessingError - I think it should be gone ultimately (in its current form). It should be a property (dataProcessingError: true) on any error (IFluidErrorBase), but it should preserve errorType if it was already IFluidErrorBase error, or add unique errorType (depends on place and condition where its thrown).
Similarly here, it should either preserve errorType (if it's our error) or put unique errorType on this object, but in all cases - add new property (let's call it SharedObjectEventListerException = true).
Basically all these qualities (is it a data processing error, is it DDS listener exception, is it something else) should compose, not overwrite one another.
Text is a bit tricky - we always try to preserve message of original error. There is no error nesting (though maybe we should add it - not for telemetry, but for real-time debugging).

What do you think about such plan?

+@markfields for his critique

Comment (ChumpChief) [reply packages/dds/shared-object-base/src/sharedObject.ts]: Yeah, maybe DataProcessingError should instead be something like UserCodeError?  Basically I'm thinking that this error is a little special because it lives on the public API boundary.  So whereas our internal errors are OK with error codes and internal knowledge about error types, this one wants to be very understandable from the outside with user-friendly messaging and visibility.

Plan is prob fine, again it will probably be easier to relax the fail-fast in the future based on feedback rather than make it more strict so starting strict is prob ok.

**Comment by vladsud** [comment]
Related thread: https://teams.microsoft.com/l/message/19:9034865f2a4f49b5ab8df4569dbb0e73@thread.skype/1637083535909?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=422665a0-7ad3-4d0d-9171-e8881d0397d9&parentMessageId=1637083535909&teamName=FFX%20(Fluid%20Framework%20and%20Experiences)&channelName=LiveSite%20%E2%9A%A0&createdTime=1637083535909

**Review by anthony-murphy — APPROVED** [review]

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 7]
```diff
@@ -108,7 +114,7 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
         protected runtime: IFluidDataStoreRuntime,
         public readonly attributes: IChannelAttributes)
     {
-        super((eventName, error) => this.runtime.raiseContainerWarning(error));
+        super((event: string | symbol, e: any) => this.eventListenerErrorHandler(event, e));
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:117]: nit: use `EventEmitterEventType` instead of `string | symbol`, right?

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 8]
```diff
@@ -126,6 +132,47 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+     * in event listeners (i.e. catch any of the issues and make determination on how to handle it).
+     * When such exceptions propagate through, most likely data model is no longer consistent, i.e.
+     * DDS state does not match what user sees. Because of it DDS moves to "corrupted state" and does not
+     * allow processing of ops or local changes, which very quickly results in container closure.
+     */
+    private eventListenerErrorHandler(event: string | symbol, e: any) {
+        const error = DataProcessingError.wrapIfUnrecognized(
+            e,
+            "SharedObjectEventListenerException",
+            undefined); // message
```
Comment (markfields) [inline packages/dds/shared-object-base/src/sharedObject.ts:169]: Probably best to not leave message undefined (just copy the error code to message field?), since we look at the error field all the time in queries.

**Review by markfields — APPROVED** [review]

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 8 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:169]: message here is ISequencedDocumentMessage. But it's optional, so I can remove this arg altogether.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 6 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts]: Let's keep this discussion (on how user errors are projected in our system) as part of https://github.com/microsoft/FluidFramework/pull/8475 and appropriate issue. It's not unique to this case as user exceptions come from various points and we need to ensure we are consistent in handling them.
Wrapping (i.e. DataProcessingError.wrapIfUnrecognized()) is the way we will handle it in all cases.
